Implement EventNotifier/EventConsumer as a generic event notification

uran0sH opened the pull request:
Introduce an abstract event notification in order to support multiple platforms. EventSender is used to send a notification, and EventReceiver is used to receive a notification. The fd used in EventSender or EventReceiver can be from eventfd, pipefd or others.

uran0sH committed:
Add generic event notification(EventNotifier/Consumer)

Introduce an abstract event notification in order to support multiple
platforms. EventNotifier is used to send a notification, and EventConsumer
is used to receive a notification. The fd used in EventNotifier or
EventConsumer can be from eventfd, pipefd or others.

Signed-off-by: Wenyu Huang <huangwenyuu@outlook.com>

uran0sH commented:
Plz review @stefano-garzarella @germag @osteffenrh

germag commented on src/event.rs:
```diff
@@ -0,0 +1,125 @@
+
+impl EventSender {
+    /// Write a value to the EventSender's fd
+    pub fn write(&self) -> result::Result<(), io::Error> {
```
as we discussed, change the name `EventSender::write()` -> `EventNotifier::notify()` (also update the commit message)

germag commented on src/event.rs:
```diff
@@ -0,0 +1,125 @@
+            Err(io::Error::last_os_error())
+        } else {
+            Ok(())
+        }
```
```rust
    let v = 1u64;
    (&self.fd).write_all(&v.to_ne_bytes())
```

germag commented on src/event.rs:
```diff
@@ -0,0 +1,125 @@
+
+/// EventReceiver: A wrapper around a file descriptor that can be used to receive events
+#[derive(Debug)]
+pub struct EventReceiver {
```
`EventReceiver::read()` to `EventConsumer::consume()`

germag commented on src/event.rs:
```diff
@@ -0,0 +1,125 @@
+            Err(io::Error::last_os_error())
+        } else {
+            Ok(buf)
+        }
```
```rust
 let mut buf = [0u8; size_of::<u64>()];
    (&self.fd).read_exact(buf.as_mut_slice()).map(|_| u64::from_ne_bytes(buf))
```

germag commented on src/event.rs:
```diff
@@ -0,0 +1,125 @@
+            })
+        }
+    }
+}
```
also add tests, you can take inspiration from those in linux/[eventfd.rs](http://eventfd.rs/)

germag reviewed (commented):
Just a general comment, please check my email, I'm doing the same review again

germag commented on src/event.rs:
```diff
@@ -0,0 +1,123 @@
+        let consumer = EventConsumer {
+            // SAFETY: Safe because we check the fd is valid. And the kernel gave us an fd that we own.
+            fd: unsafe { File::from_raw_fd(reader.into_raw_fd()) },
+        };
```
why not?
```rust
        // SAFETY: Safe because we check the fd is valid. And the kernel gave us an fd that we own.
        let notifier = unsafe { EventNotifier::from_raw_fd(writer.into_raw_fd()) };
        
        // SAFETY: Safe because we check the fd is valid. And the kernel gave us an fd that we own.
        let consumer = unsafe { EventConsumer::from_raw_fd(reader.into_raw_fd()) };
```

and same below

germag commented on src/event.rs:
```diff
@@ -0,0 +1,123 @@
+        let consumer = EventConsumer {
+            // SAFETY: Safe because we check the fd is valid. And the kernel gave us an fd that we own.
+            fd: unsafe { File::from_raw_fd(reader.into_raw_fd()) },
+        };
```
same as before, use `from_raw_fd()`

germag commented on src/event.rs:
```diff
@@ -0,0 +1,123 @@
+        };
+
+        let cloned_notifier = notifier.try_clone().unwrap();
+        let cloned_consumer = consumer.try_clone().unwrap();
```
unwrap() -> expect()

germag reviewed (commented):
I think it looks good

dependabot[bot] committed:
build(deps): bump rust-vmm-ci from `c3ef897` to `aa2f8a9`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `c3ef897` to `aa2f8a9`.
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/c3ef8977388ccddb8c68b106737fa2c674b0d87a...aa2f8a90e46533de8a58f19cd0c417de856d4437)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-version: aa2f8a90e46533de8a58f19cd0c417de856d4437
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

roypat committed:
chore: appease clippy

Signed-off-by: Patrick Roy <roypat@amazon.co.uk>

stefano-garzarella commented:
Can you close this PR? I'm receiving duplicated notification for this and upstream one. Please check also other PRs. If you have a one in your private fork that now is also upstream, close the private one.
